crates/accelerate/src/unitary_compose.rs

```diff
@@ -0,0 +1,149 @@
+        if front {
+            return gate_unitary.dot(overall_unitary);
+        } else {
+            return overall_unitary.dot(gate_unitary);
+        }
+    }
+    // Compose with other on subsystem
+    let num_indices = gate_qubits;
+    let shift = if front { gate_qubits } else { 0usize };
+    let right_mul = front;
+
+    //Reshape current matrix
+    //Note that we must reverse the subsystem dimension order as
+    //qubit 0 corresponds to the right-most position in the tensor
+    //product, which is the last tensor wire index.
+    let tensor = per_qubit_shaped(gate_unitary);
+    let mat = per_qubit_shaped(overall_unitary);
+    let indices = qubits
+        .iter()
+        .map(|q| num_indices - 1 - q)
+        .collect::<Vec<usize>>();
+    let num_rows = usize::pow(2, num_indices as u32);
+
+    _einsum_matmul(&tensor, &mat, &indices, shift, right_mul)
+        .as_standard_layout()
+        .into_shape((num_rows, num_rows))
+        .unwrap()
+        .into_dimensionality::<ndarray::Ix2>()
+        .unwrap()
+        .to_owned()
+}
+
+// Reshape an input matrix to (2, 2, ..., 2) depending on its dimensionality
+fn per_qubit_shaped<'a>(array: &ArrayView2<'a, Complex<f64>>) -> ArrayView<'a, Complex64, IxDyn> {
+    let overall_shape = (0..array.shape()[0].ilog2() as usize)
+        .flat_map(|_| [2, 2])
+        .collect::<Vec<usize>>();
+    array.into_shape(overall_shape).unwrap()
+}
+
+// Determine einsum strings for perform a matrix multiplication on the input matrices
+fn _einsum_matmul(
+    tensor: &ArrayView<Complex64, IxDyn>,
+    mat: &ArrayView<Complex64, IxDyn>,
+    indices: &Vec<usize>,
+    shift: usize,
+    right_mul: bool,
+) -> Array<Complex64, IxDyn> {
+    let rank = tensor.ndim();
+    let rank_mat = mat.ndim();
+    if rank_mat % 2 != 0 {
+        panic!("Contracted matrix must have an even number of indices.");
+    }
+    // Get einsum indices for tensor
+    let mut indices_tensor = (0..rank).collect::<Vec<usize>>();
+    for (j, index) in indices.iter().enumerate() {
+        indices_tensor[index + shift] = rank + j;
+    }
+    // Get einsum indices for mat
+    let mat_contract = (rank..rank + indices.len()).rev().collect::<Vec<usize>>();
+    let mat_free = indices
+        .iter()
+        .rev()
+        .map(|index| index + shift)
+        .collect::<Vec<usize>>();
+    let indices_mat = if right_mul {
+        [mat_contract, mat_free].concat()
+    } else {
+        [mat_free, mat_contract].concat()
+    };
+    let tensor_einsum: String = unsafe {
+        String::from_utf8_unchecked(indices_tensor.iter().map(|c| LOWERCASE[*c]).collect())
+    };
+    let mat_einsum: String =
+        unsafe { String::from_utf8_unchecked(indices_mat.iter().map(|c| LOWERCASE[*c]).collect()) };
+    einsum(
+        format!("{},{}", tensor_einsum, mat_einsum).as_str(),
+        &[tensor, mat],
+    )
+    .unwrap()
+}
+
+fn _einsum_matmul_helper(qubits: &[usize], num_qubits: usize) -> [String; 4] {
```

Comment:
Use `u32` for Qubit indices here as well:
```suggestion
fn _einsum_matmul_helper(qubits: &[u32], num_qubits: usize) -> [String; 4] {
```